src/MongoDB.Driver/Linq/Processors/ExpressionValidator.cs

```diff
@@ -0,0 +1,122 @@
+            }
+
+            validator.Visit(expression);
+        }
+
+        /// <inheritdoc />
+        protected override Expression VisitMethodCall(MethodCallExpression node)
+        {
+            if (node.Arguments.Any())
+            {
+                foreach (var argument in node.Arguments.Where(c => !(c is ConstantExpression || c is ParameterExpression))) // do not check the bottom of tree branches.
+                {
+                    var usedParameters = ParameterExpressionCollector.Collect(argument).ToList();
+
+                    if (!usedParameters.Any())
+                    {
+                        continue;
+                    }
+
+                    var memberExpression = argument as MemberExpression;
+                    if (memberExpression != null)
+                    {
+                        // add top-level expression parameters to the list
+                        // with parent's parameters.
+                        // NOTE: these parameters will be added to all members,
+                        // even if the member is not in the top-level, but since
+                        // top-level parameters should not be placed in any level 
+                        // except of direct children, it's not a problem
+                        AddTopLevelExpressionParameters(_inParametersScope);
+                        continue;
+                    }
+
+                    foreach (var usedParameter in usedParameters)
+                    {
+                        // check the current LINQ expression on whether
+                        // the parent expression parameters are used
+                        // on the current expression level or not.
+                        if (_inParametersScope.Contains(usedParameter))
+                        {
+                            throw new NotSupportedException($"The LINQ expression: {argument} has the member \"{usedParameter}\" which can not be used to build a correct Mongo query");
+                        }
+                    }
+
+                    var methodExpression = argument as MethodCallExpression;
+                    if (methodExpression != null)
+                    {
+                        Visit(methodExpression);
+                    }
+                    else
+                    {
+                        // save the parent expression parameters for the current level
+                        // to restore their values after visiting of child expression nodes.
+                        var storedParentParameters = _inParametersScope;
+                        var lambdaExpression = argument as LambdaExpression;
+                        if (lambdaExpression != null)
+                        {
+                            _inParametersScope = lambdaExpression.Parameters.Select(c => c.Name).ToList();
+                        }
+
+                        base.Visit(argument);
```

Comment:
I don't think "base" is needed here. Let's remove it if it is not needed, because it's presence implies that it is needed and that just causes confusion.

Reply:
`Base` is needed here, since it allows us to visit one level down.